crates/client/state-sync/src/lib.rs

```diff
@@ -0,0 +1,461 @@
+                    }
+                    if eth_from_height == highest_block_number {
+                        tokio::time::sleep(Duration::from_secs(5)).await;
+                    }
+                }
+                Err(e) => {
+                    error!(target: LOG_TARGET, "get highest block number from l1 has error {:#?}", e);
+                }
+            };
+        }
+    };
+
+    let state_write_task = async move {
+        loop {
+            if let Some(fetched_states) = rx.next().await {
+                for state in fetched_states.iter() {
+                    let _ = state_writer.apply_state_diff(
+                        state.l1_l2_block_mapping.l2_block_number,
+                        state.l1_l2_block_mapping.l2_block_hash,
+                        &state.state_diff,
+                    );
+                }
+
+                if let Some(last) = fetched_states.last() {
+                    if let Err(e) = madara_backend.meta().write_last_l1_l2_mapping(&last.l1_l2_block_mapping) {
+                        error!(target: LOG_TARGET, "write to madara backend has error {:#?}", e);
+                        break;
+                    }
+                }
+            }
+        }
+    };
+
+    future::select(Box::pin(fetcher_task), Box::pin(state_write_task)).map(|_| ())
+}
+
+/// Converts a U256 value to an H256 value.
+///
+/// This helper function takes a U256 value and converts it to an H256 value by
+/// converting the U256 value to a big-endian byte array, then copying the first
+/// 32 bytes into an H256 array to obtain the H256 value.
+///
+/// # Arguments
+///
+/// * `u256` - The U256 value to be converted to H256.
+///
+/// # Returns
+///
+/// An H256 value converted from the provided U256 value.
+fn u256_to_h256(u256: U256) -> H256 {
+    let mut bytes = [0; 32];
+    u256.to_big_endian(&mut bytes);
+    let mut h256_bytes = [0; 32];
+    h256_bytes.copy_from_slice(&bytes[..32]);
+    H256::from(h256_bytes)
+}
+
+/// Represents various error types that can occur during state synchronization or interaction with
+/// L1/L2 chains.
+#[derive(Debug, Clone)]
+pub enum Error {
+    /// Error indicating that the data is already present in the chain.
+    AlreadyInChain,
+    /// Error indicating an unknown block or data on the chain.
+    UnknownBlock,
+    /// Error occurring during transaction construction with a specific message.
+    ConstructTransaction(String),
+    /// Error while committing data to storage with a specific message.
+    CommitStorage(String),
+    /// Error related to connection issues with L1 chain with a specific message.
+    L1Connection(String),
+    /// Error decoding an event from L1.
+    L1EventDecode,
+    /// Error related to state handling on L1 with a specific message.
+    L1StateError(String),
+    /// Error due to a type mismatch or inconsistency with a specific message.
+    TypeError(String),
+    /// Any other unspecified error with a specific message.
+    Other(String),
```

Comment:
Don't use a `String` as a way to represent the original error. 
Use the error type itself.
You can achieve this easily with `thiserror` and their `#[from]` macro

Reply:
I have rewritten the `error` using `thiserror`.